hazelcast/src/hazelcast/client/serialization/pimpl/SerializationService.cpp

```diff
@@ -424,278 +420,72 @@ namespace hazelcast {
-
-                    return dataInput.readLongArray();
-                }
-
-                template<>
-                std::auto_ptr< std::vector<float> >  SerializationService::toObject(const Data &data) {
-                    CHECK_NULL(std::vector<float>);
-
-                    DataInput dataInput(data.toByteArray(), Data::DATA_OFFSET);
-
-                    int32_t typeId = data.getType();
-
-                    constants.checkClassType(SerializationConstants::CONSTANT_TYPE_FLOAT_ARRAY, typeId);
-
-                    return dataInput.readFloatArray();
+                std::auto_ptr<Data> SerializationService::toObject(const Data *data) {
+                    return std::auto_ptr<Data>((Data *) data);
                 }
 
-                template<>
-                std::auto_ptr<std::vector<double> > SerializationService::toObject(const Data &data) {
-                    CHECK_NULL(std::vector<double>);
-
-                    DataInput dataInput(data.toByteArray(), Data::DATA_OFFSET);
-
-                    int32_t typeId = data.getType();
-
-                    constants.checkClassType(SerializationConstants::CONSTANT_TYPE_DOUBLE_ARRAY, typeId);
-
-                    return dataInput.readDoubleArray();
+                const byte SerializationService::getVersion() const {
+                    return 1;
                 }
 
-                template<>
-                std::auto_ptr<std::string> SerializationService::toObject(const Data &data) {
-                    CHECK_NULL(std::string);
-
-                    DataInput dataInput(data.toByteArray(), Data::DATA_OFFSET);
-
-                    int32_t typeId = data.getType();
+                ObjectType SerializationService::getObjectType(const Data *data) {
+                    ObjectType type;
 
-                    constants.checkClassType(SerializationConstants::CONSTANT_TYPE_STRING, typeId);
+                    if (NULL == data) {
+                        return type;
+                    }
 
-                    return dataInput.readUTF();
-                }
+                    type.typeId = data->getType();
 
-                template<>
-                std::auto_ptr<std::vector<std::string> > SerializationService::toObject(const Data &data) {
-                    CHECK_NULL(std::vector<std::string>);
+                    // Constant 4 is Data::TYPE_OFFSET. Windows DLL export does not
+                    // let usage of static member.
+                    DataInput dataInput(data->toByteArray(), 4);
 
-                    DataInput dataInput(data.toByteArray(), Data::DATA_OFFSET);
+                    ObjectDataInput objectDataInput(dataInput, getSerializerHolder());
 
-                    int32_t typeId = data.getType();
+                    if (SerializationConstants::CONSTANT_TYPE_DATA == type.typeId ||
+                            SerializationConstants::CONSTANT_TYPE_PORTABLE == type.typeId) {
+                        assert(type.typeId == objectDataInput.readInt());
```

Comment:
Be careful here. You need to get this readInt out of assert. In `run` mode, readInt will not be called.